staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1059 @@
+	remoteAttRef := ca.fromAttToEnquableRef(remoteAtt)
+	ca.queue.Add(remoteAttRef)
+}
+
+func (ca *ConnectionAgent) getLocalAttIfc(ifcOwnerAttNSN k8stypes.NamespacedName) (ifc netfabric.NetworkInterface, ifcFound bool) {
+	ca.localAttIfcsMutex.RLock()
+	defer ca.localAttIfcsMutex.RUnlock()
+	ifc, ifcFound = ca.localAttIfcs[ifcOwnerAttNSN]
+	return
+}
+
+func (ca *ConnectionAgent) getRemoteAttIfc(ifcOwnerAttNSN k8stypes.NamespacedName) (ifc netfabric.NetworkInterface, ifcFound bool) {
+	ca.remoteAttIfcsMutex.RLock()
+	defer ca.remoteAttIfcsMutex.RUnlock()
+	ifc, ifcFound = ca.remoteAttIfcs[ifcOwnerAttNSN]
+	return
+}
+
+func (ca *ConnectionAgent) testAndSwapLocalAttIfcIfVNIsMatch(ifcOwnerAttNSN k8stypes.NamespacedName,
+	newIfc netfabric.NetworkInterface) (oldIfc netfabric.NetworkInterface, oldIfcExists, newIfcSet bool) {
+
+	ca.localAttIfcsMutex.Lock()
+	defer ca.localAttIfcsMutex.Unlock()
+	oldIfc, oldIfcExists = ca.localAttIfcs[ifcOwnerAttNSN]
+	if oldIfcExists && oldIfc.VNI != newIfc.VNI {
+		newIfcSet = false
+		return
+	}
+	ca.localAttIfcs[ifcOwnerAttNSN] = newIfc
+	newIfcSet = true
+	return
+}
+
+func (ca *ConnectionAgent) testAndSwapRemoteAttIfcIfVNIsMatch(ifcOwnerAttNSN k8stypes.NamespacedName,
+	newIfc netfabric.NetworkInterface) (oldIfc netfabric.NetworkInterface, oldIfcExists, newIfcSet bool) {
+
+	ca.remoteAttIfcsMutex.Lock()
+	defer ca.remoteAttIfcsMutex.Unlock()
+	oldIfc, oldIfcExists = ca.remoteAttIfcs[ifcOwnerAttNSN]
+	if oldIfcExists && oldIfc.VNI != newIfc.VNI {
+		newIfcSet = false
+		return
+	}
+	ca.remoteAttIfcs[ifcOwnerAttNSN] = newIfc
+	newIfcSet = true
+	return
+}
+
+func (ca *ConnectionAgent) removeLocalAttIfc(ifcOwnerAttNSN k8stypes.NamespacedName) {
+	ca.localAttIfcsMutex.Lock()
+	defer ca.localAttIfcsMutex.Unlock()
+	delete(ca.localAttIfcs, ifcOwnerAttNSN)
+}
+
+func (ca *ConnectionAgent) removeRemoteAttIfc(ifcOwnerAttNSN k8stypes.NamespacedName) {
+	ca.remoteAttIfcsMutex.Lock()
+	defer ca.remoteAttIfcsMutex.Unlock()
+	delete(ca.remoteAttIfcs, ifcOwnerAttNSN)
+}
+
+func (ca *ConnectionAgent) fromAttToEnquableRef(att *netv1a1.NetworkAttachment) (enquableRef attQueueRef) {
+	enquableRef.nsn = attNSN(att)
+	enquableRef.vni = att.Spec.VNI
+	if att.Spec.Node == ca.localNodeName {
+		enquableRef.local = true
+	} else {
+		enquableRef.local = false
+	}
+	return
+}
+
+// Return a string representing a field selector that matches NetworkAttachments
+// that run on the local node and have a virtual IP.
+func (ca *ConnectionAgent) localAttWithAnIPSelector() string {
+	// localAttSelector expresses the constraint that the NetworkAttachment runs on this node.
+	localAttSelector := attNodeFieldName + equal + ca.localNodeName
+
+	// Express the constraint that the NetworkAttachment has a virtual IP by saying
+	// that the field containig the virtual IP must not be equal to the empty string.
+	attWithAnIPSelector := attIPFieldName + notEqual
```

Comment:
Is this the right way to write the empty string, or must it be quoted?

Reply:
The correct way is the one in the code, with no quotation.